### PR TITLE
Tickets/dm 34991

### PR DESCRIPTION
**Pull request by tribeiro**
The idea of this PR is to add a new script to execute the optical alignment of AuxTel using the wep code we will also use for the main telescope. Because the script is doing the exact same thing as the currently existing `LatissCWFSAlign`, and in order to avoid copy/pasting a bunch of code I decided to create a base class `LatissBaseAlign` that contains all (or most of) the common code. This includes the script configuration, slewing/offsetting/taking data etc. Then there is an abstract method called `run_align` which is in charge of running the wavefront estimation code and outputting a result.

The original class was updated to use the new base class.

A new class `LatissWEPAlign` was added (alongside a script to run it), which implements the wep part.

Some things to note:

1 - This new class runs the code in a separate process, rather than a separate thread. This keeps he main thread free to manage the script. The caveat to that is that we need to isolate all the data going to and coming back from the process. To that, I am not pulling all the data generated on the process back, only what is really needed, otherwise we would loose a lot of time with Process IO.

2 - It is using BestEffortISR to process the data and `QuickFrame` to find the sources, just like the original script. I decided for this venue to reduce the adoption friction, since we have been testing this for quite a while successfully. 

3 - I could not find a good way to reuse the unit tests. So here I had to resort to copy/paste. :(

4 - Speaking of unit tests, they are not passing. The reason, as you saw the discussion on slack, is that the results we get from cwfs and wep are not compatible within the specified tolerances. Then we found the issue with the obscuration in the original version. I guess we can think about what we will do in the future. Note that these tests only run on the test stand and you need to have ts_wep for the wep tests to run. So, the CI tests passes, but only because it is skipping the tests that need data.

### CODE REVIEW

**Comment by tribeiro** [comment]
Just a note on the test duplication.. I found a way to avoid duplicating it using this library https://pypi.org/project/parameterized/ but it is not part of our stack. :(

File: python/lsst/ts/externalscripts/auxtel/latiss_base_align.py [hunk 1]
```diff
@@ -0,0 +1,788 @@
...
+    zernikes_rot: typing.Tuple[float, float, float]
+    offset_hex: typing.Tuple[float, float, float]
+    offset_tel: typing.Tuple[float, float, float]
+
+
+class LatissBaseAlign(salobj.BaseScript, metaclass=abc.ABCMeta):
+    """Implements generic behavior for scripts that execute curvature wavefront
+    sensing, abstracting the part that performs the measurements.
+
+    This meta Script implements the basic functionality of finding targets and
```
Comment (patrickingraham) [inline python/lsst/ts/externalscripts/auxtel/latiss_base_align.py:54]: what is a "meta script" ?

File: python/lsst/ts/externalscripts/auxtel/latiss_base_align.py [hunk 2]
```diff
@@ -0,0 +1,788 @@
...
+    taking the (intra/extra focal) data. Child classes are left to implement
```
Comment (patrickingraham) [inline python/lsst/ts/externalscripts/auxtel/latiss_base_align.py:55]: As a generic comment, should we not be using semantic line versioning for all comments/docstrings etc?

File: python/lsst/ts/externalscripts/auxtel/latiss_base_align.py [hunk 3]
```diff
@@ -0,0 +1,788 @@
...
+
+        # Timeouts used for telescope commands
+        self.timeout_short = 5.0  # used with hexapod offset command
+        self.timeout_long = 30.0  # used to wait for in-position event from hexapod
+        # Have discovered that the occasional image will take 12+ seconds
+        # to ingest
+        self.timeout_get_image = 20.0
+        self.data_pool_sleep = 5.0
+
+        # Sensitivity matrix: mm of hexapod motion for nm of wfs. To figure out
```
Comment (patrickingraham) [inline python/lsst/ts/externalscripts/auxtel/latiss_base_align.py:117]: should probably reference the technote here.
https://tstn-016.lsst.io/

File: python/lsst/ts/externalscripts/auxtel/latiss_base_align.py [hunk 4]
```diff
@@ -0,0 +1,788 @@
...
+            [
+                [np.cos(np.radians(angle)), -np.sin(np.radians(angle)), 0.0],
+                [np.sin(np.radians(angle)), np.cos(np.radians(angle)), 0.0],
+                [0.0, 0.0, 1.0],
+            ]
+        )
+
+        # Matrix to map hexapod offset to alt/az offset in the focal plane
+        # units are arcsec/mm. X-axis is Elevation
+        # Measured with data from AT run SUMMIT-5027, still unverified.
```
Comment (patrickingraham) [inline python/lsst/ts/externalscripts/auxtel/latiss_base_align.py:138]: I don't think we have a reason to believe this is incorrect?

File: python/lsst/ts/externalscripts/auxtel/latiss_base_align.py [hunk 5]
```diff
@@ -0,0 +1,788 @@
...
+        results : `LatissAlignResults`
+            Results of the wavefront sensing.
+        """
+        rot_zern = np.matmul(
+            self.zern, self.matrix_rotation(self.angle + self.camera_rotation_angle)
+        )
+        hexapod_offset = np.matmul(rot_zern, self.matrix_sensitivity)
+        tel_offset = np.matmul(hexapod_offset, self.hexapod_offset_scale)
+
+        self.log.info(
```
Comment (patrickingraham) [inline python/lsst/ts/externalscripts/auxtel/latiss_base_align.py:352]: I really wish we could publish these results in a formatted way... like write an object to the LFA...
How does the MTAOS do it?
Can we be writing a file here?

File: python/lsst/ts/externalscripts/auxtel/latiss_base_align.py [hunk 6]
```diff
@@ -0,0 +1,788 @@
...
+                )
+                if checkpoint:
+                    await self.checkpoint(
+                        f"[{i + 1}/{self.max_iter}]: CWFS focus error too large."
+                    )
+                await self.hexapod_offset(focus_offset / 2.0)
+            else:
+                self.offset_total_focus += focus_offset
+                self.log.info(
+                    f"Applying offset: x={coma_x}, y={coma_y}, z={focus_offset}."
```
Comment (patrickingraham) [inline python/lsst/ts/externalscripts/auxtel/latiss_base_align.py:755]: maybe constrain the values here such as...
`{coma_x:0.3f}`

File: python/lsst/ts/externalscripts/auxtel/latiss_base_align.py [hunk 7]
```diff
@@ -0,0 +1,788 @@
...
+                    same size as the intra. This value is always positive to compensate for
+                    the change in magnification from moving the secondary mirror.
+                type: number
+                default: 0.0011
+              large_defocus:
+                description: >-
+                    Defines a large defocus. If the measured defocus is larger than this value,
+                    apply only half of correction.
+                type: number
+                default: 0.08
```
Comment (patrickingraham) [inline python/lsst/ts/externalscripts/auxtel/latiss_base_align.py:478]: I think we can increase this now... maybe to 0.12?

File: python/lsst/ts/externalscripts/auxtel/latiss_base_align.py [hunk 8]
```diff
@@ -0,0 +1,788 @@
...
+              threshold:
+                 description: >-
+                   Focus correction threshold. If correction is lower than this
+                   value, stop correction loop.
+                 type: number
+                 default: 0.01
```
Comment (patrickingraham) [inline python/lsst/ts/externalscripts/auxtel/latiss_base_align.py:484]: Here too, we can boost it to 0.015

File: python/lsst/ts/externalscripts/auxtel/latiss_base_align.py [hunk 9]
```diff
@@ -0,0 +1,788 @@
...
+        can be set to False, which allows one to mock the remotes behaviour.
+    descr : `str`, optional
+        Short description of the script.
+
+    Notes
+    -----
+    **Checkpoints**
+
+    - "Slewing to ...": Before slewing to target.
+    - "Detection image": If taking in-focus detection image.
```
Comment (patrickingraham) [inline python/lsst/ts/externalscripts/auxtel/latiss_base_align.py:73]: what are we detecting? See discussion about this in the config area

File: python/lsst/ts/externalscripts/auxtel/latiss_base_align.py [hunk 10]
```diff
@@ -0,0 +1,788 @@
...
+                default: empty_1
+              exposure_time:
+                description: The exposure time to use when taking intra/extra focal images (sec).
+                type: number
+                default: 30.
+              acq_exposure_time:
+                description: The exposure time to use when taking an in focus acquisition image(sec).
+                type: number
+                default: 5.
+              take_detection_image:
```
Comment (patrickingraham) [inline python/lsst/ts/externalscripts/auxtel/latiss_base_align.py:456]: So this was used for the WEP testing parts before, correct? Is it still required? 
If so, maybe add a sentence in the description saying why it exists and when to use it.

File: python/lsst/ts/externalscripts/auxtel/latiss_base_align.py [hunk 11]
```diff
@@ -0,0 +1,788 @@
...
+
+        Parameters
+        ----------
+        metadata : `salobj.type_hints.BaseMsgType`
+            Script metadata topic. The information is set on the topic
+            directly.
+        """
+        # It takes about 300s to run the cwfs code, plus the two acquisition
+        # exposures.
+        metadata.duration = 300.0 + 2.0 * self.exposure_time
```
Comment (patrickingraham) [inline python/lsst/ts/externalscripts/auxtel/latiss_base_align.py:593]: I think we can update this to be max_iterations * exposure_time*2 + detection image + in-focus image time etc. Plus a small fudge factor of course.

File: python/lsst/ts/externalscripts/auxtel/latiss_base_align.py [hunk 12]
```diff
@@ -0,0 +1,788 @@
...
+        RuntimeError:
+            If coordinates are malformed.
+        """
+
+        if self.cwfs_target is not None and self.cwfs_target_dec is None:
+            if checkpoint:
+                await self.checkpoint(f"Slewing to object: {self.cwfs_target}")
+            await self.atcs.slew_object(
+                name=self.cwfs_target,
+                rot=0.0,
```
Comment (patrickingraham) [inline python/lsst/ts/externalscripts/auxtel/latiss_base_align.py:616]: Do we need to do this at rot=0 ? It adds overhead, especially when using the parallactic rotType.

File: python/lsst/ts/externalscripts/auxtel/latiss_base_align.py [hunk 13]
```diff
@@ -0,0 +1,788 @@
...
+                await self.checkpoint(
+                    f"Slewing to icrs coordinates: {self.cwfs_target} @ "
+                    f"ra/dec = {self.cwfs_target_ra}/{self.cwfs_target_dec}."
+                )
+
+            await self.atcs.slew_icrs(
+                ra=self.cwfs_target_ra,
+                dec=self.cwfs_target_dec,
+                target_name=self.cwfs_target,
+                rot=0.0,
```
Comment (patrickingraham) [inline python/lsst/ts/externalscripts/auxtel/latiss_base_align.py:636]: same deal as above

File: python/lsst/ts/externalscripts/auxtel/latiss_base_align.py [hunk 14]
```diff
@@ -0,0 +1,788 @@
...
+                self.offset_total_coma_x += coma_x
+                self.offset_total_coma_y += coma_y
+                await self.hexapod_offset(focus_offset, x=coma_x, y=coma_y)
+                if self.offset_telescope:
+                    tel_el_offset, tel_az_offset = (
+                        results.offset_tel[0],
+                        results.offset_tel[1],
+                    )
+                    self.log.info(
+                        f"Applying telescope offset az/el: {tel_az_offset}/{tel_el_offset}."
```
Comment (patrickingraham) [inline python/lsst/ts/externalscripts/auxtel/latiss_base_align.py:770]: `{tel_az_offset:0.3f}`

File: python/lsst/ts/externalscripts/auxtel/latiss_cwfs_align.py [hunk 15]
```diff
@@ -204,43 +108,20 @@ def __init__(self, index=1, remotes=True):
...
-        return self._dz
-
-    @property
-    def binning(self):
+    def binning(self) -> int:
         return self._binning
 
     @binning.setter
-    def binning(self, value):
+    def binning(self, value: int) -> None:
```
Comment (patrickingraham) [inline python/lsst/ts/externalscripts/auxtel/latiss_cwfs_align.py:120]: Nice to see this!

File: python/lsst/ts/externalscripts/auxtel/latiss_cwfs_align.py [hunk 16]
```diff
@@ -265,128 +146,13 @@ def dz(self, value):
...
-                or check_ss2 is True
-            ):
-                check_ss2 = True
-                if state.substate == 0:
-                    break
-
-    async def run_cwfs(self):
-        """Runs CWFS code on intra/extra focal images.
+    async def run_align(self) -> LatissAlignResults:
+        """Runs curvature wavefront sensing code.
```
Comment (patrickingraham) [inline python/lsst/ts/externalscripts/auxtel/latiss_cwfs_align.py:150]: Maybe we should expand this to emphasize it's the original code and not the WEP code?

Runs CWFS analysis using the ts_cwfs package.

Comment (patrickingraham) [reply python/lsst/ts/externalscripts/auxtel/latiss_cwfs_align.py:150]: Somewhere a log info message should also be written that says which analysis is being used.

**Review by patrickingraham — CHANGES_REQUESTED** [review]
First round of comments, most are really minor... but might be worth addressing before I look at the WEP part.

File: python/lsst/ts/externalscripts/auxtel/latiss_base_align.py (hunk 1 above)
Comment (tribeiro) [reply python/lsst/ts/externalscripts/auxtel/latiss_base_align.py:54]: I guess, I made the term up, But basically, it is a "meta class" that implements a SAL Script hence, a "meta script". In essence, this is not a script that you can use as-is, it needs specializations to implement the core methods that will actually do the wavefront sensing.

File: python/lsst/ts/externalscripts/auxtel/latiss_base_align.py (hunk 2 above)
Comment (tribeiro) [reply python/lsst/ts/externalscripts/auxtel/latiss_base_align.py:55]: Uhm, not sure I understand what you mean. Can you rephrase the question?

File: python/lsst/ts/externalscripts/auxtel/latiss_base_align.py (hunk 5 above)
Comment (tribeiro) [reply python/lsst/ts/externalscripts/auxtel/latiss_base_align.py:352]: The MTAOS has events for these, but the results are also stored in the butler by the pipeline task. We can certainly publish them to the FLA tough.

I can make a ticket for that, don't want to make conflate too many changes in one go.

File: python/lsst/ts/externalscripts/auxtel/latiss_base_align.py (hunk 10 above)
Comment (tribeiro) [reply python/lsst/ts/externalscripts/auxtel/latiss_base_align.py:456]: Yes, I added this for early WEP testing. In my implementation I ended up using the same "ISR" and detection code we use for the old cwfs code so, it is not used for processing the data. I though about leaving it just because it might be a nice debugging tool eventually. But it is right that it adds unnecessary complexity, so maybe I should remove it. 

I'm not sure, what do you think? remove or keep it for debugging purposes and add a note about it?

File: python/lsst/ts/externalscripts/auxtel/latiss_base_align.py (hunk 12 above)
Comment (tribeiro) [reply python/lsst/ts/externalscripts/auxtel/latiss_base_align.py:616]: We don't need to do it at `rot=0.0` what do you suggest tough? should we add this as a parameter? or maybe try to get the current rotator position and use it in `PhysicalSky` mode?

File: python/lsst/ts/externalscripts/auxtel/latiss_cwfs_align.py (hunk 16 above)
Comment (tribeiro) [reply python/lsst/ts/externalscripts/auxtel/latiss_cwfs_align.py:150]: I can make this information part of the script description which appears on LOVE at a special/static location.

**Comment by tribeiro** [comment]
@patrickingraham , I think I addressed al your current comments. Some of them I think I need some input to resolve.

File: python/lsst/ts/externalscripts/auxtel/latiss_base_align.py (hunk 10 above)
Comment (patrickingraham) [reply python/lsst/ts/externalscripts/auxtel/latiss_base_align.py:456]: I think you can leave it, just add a sentence to the description. 
Something like, "Used for early WEP testing but maintained for possible debugging scenarios"

File: python/lsst/ts/externalscripts/auxtel/latiss_base_align.py (hunk 12 above)
Comment (patrickingraham) [reply python/lsst/ts/externalscripts/auxtel/latiss_base_align.py:634]: why not make the `rot` and `RotType` config parameters? Default is SkyAuto (so basically do whatever is fastest).

Feeding the current position to PhysicalSky seems like we're adding complication to simplify standard behavior?

File: python/lsst/ts/externalscripts/auxtel/latiss_cwfs_align.py (hunk 16 above)
Comment (patrickingraham) [reply python/lsst/ts/externalscripts/auxtel/latiss_cwfs_align.py]: Ok, that'd be nice. But it goes into the log as well? Just trying to see how one would figure this out when mining back through old data.

**Review by patrickingraham — COMMENTED** [review]
Sent replies to comments, I'll look at the WEP part shortly...

File: python/lsst/ts/externalscripts/auxtel/latiss_base_align.py (hunk 2 above)
Comment (tribeiro) [reply python/lsst/ts/externalscripts/auxtel/latiss_base_align.py:55]: uhm.. I'm not sure that would work for docstrings though... Because we have limits on the line lengths and such.

File: python/lsst/ts/externalscripts/auxtel/latiss_cwfs_align.py (hunk 16 above)
Comment (tribeiro) [reply python/lsst/ts/externalscripts/auxtel/latiss_cwfs_align.py]: So, instead of going to logMessage, this information would appear in the `Script.logevent_description` topic.. which I think might even make it easier to track down...

image

File: python/lsst/ts/externalscripts/auxtel/latiss_base_align.py (hunk 2 above)
Comment (patrickingraham) [reply python/lsst/ts/externalscripts/auxtel/latiss_base_align.py:55]: Not worth holding this up, but shouldn't it be done anyways to the extent possible? Meaning a new sentence goes on a new line? It should reduce the amount of differences when making text changes and even help avoid passing limits.

File: python/lsst/ts/externalscripts/auxtel/latiss_wep_align.py [hunk 17]
```diff
@@ -0,0 +1,286 @@
...
+        result_extra.brightestObjCentroidCofM[0]
+        - result_intra.brightestObjCentroidCofM[0]
+    )
+    dx = (
+        result_extra.brightestObjCentroidCofM[1]
+        - result_intra.brightestObjCentroidCofM[1]
+    )
+    dr = np.sqrt(dy**2 + dx**2)
+
+    position_out_of_range = dr > 100
```
Comment (patrickingraham) [inline python/lsst/ts/externalscripts/auxtel/latiss_wep_align.py:180]: I don't like that this is hardcoded... I know i'm the one that did it... I have mixed feelings about making it a config parameter.

In fact, since the entire catalogue is being used (below), can we just get rid of it?

File: python/lsst/ts/externalscripts/auxtel/latiss_cwfs_align.py (hunk 16 above)
Comment (tribeiro) [reply python/lsst/ts/externalscripts/auxtel/latiss_cwfs_align.py]: Not at all..

File: python/lsst/ts/externalscripts/auxtel/latiss_wep_align.py [hunk 18]
```diff
@@ -0,0 +1,286 @@
...
+    config.donutStampSize = donut_diameter
+    config.donutTemplateSize = donut_diameter
+    config.opticalModel = "onAxis"
+
+    task = EstimateZernikesLatissTask(config=config)
+
+    camera = best_effort_isr.butler.get(
+        "camera",
+        dataId={"instrument": "LATISS"},
+        collections="LATISS/calib/unbounded",
```
Comment (patrickingraham) [inline python/lsst/ts/externalscripts/auxtel/latiss_wep_align.py:201]: what is this collection?
I guess i'm a little confused what is going on here. This grabs information about the camera that is needed by the  `EstimateZernikesLatissTask` method?
The reason you don't just put it inside the class is why? Speed? The class is already latiss specific...

File: tests/auxtel/test_latiss_cwfs_align.py [hunk 19]
```diff
@@ -431,32 +430,32 @@ async def test_take_sequence(self):
...
-                f"Hex_tol is {np.matmul(zern_tol, self.script.sensitivity_matrix)}"
+                f"Hex_tol is {np.matmul(zern_tol, self.script.matrix_sensitivity)}"
             )
 
             tel_offset_tol = np.matmul(hex_tol, self.script.hexapod_offset_scale)
             logger.debug(f"tel_offset_tol is {tel_offset_tol}")
 
             results = self.script.calculate_results()
-            for i, z in enumerate(results["zerns"]):
+            for i, z in enumerate(results.zernikes):
```
Comment (patrickingraham) [inline tests/auxtel/test_latiss_cwfs_align.py:442]: I definitely like the object better than the dictionary.

File: tests/auxtel/test_latiss_wep_align.py [hunk 20]
```diff
@@ -0,0 +1,676 @@
...
+import pytest
+
+try:
+    # TODO: (DM-24904) Remove this try/except clause when WEP is adopted
+    from lsst.ts import wep
+
+    WEP_AVAILABLE = True
+except ImportError:
+    WEP_AVAILABLE = False
+    warnings.warn("Could not import cwfs package. Most tests will be skipped.")
```
Comment (patrickingraham) [inline tests/auxtel/test_latiss_wep_align.py:38]: wep not cwfs

File: tests/auxtel/test_latiss_wep_align.py [hunk 21]
```diff
@@ -0,0 +1,676 @@
...
+# Make obslsst translators less chatty
+logging.getLogger("lsst.obs.lsst.translators").setLevel(logging.WARNING)
+logging.getLogger("astro_metadata_translator").setLevel(logging.WARNING)
+logging.getLogger("flake8.style_guide").setLevel(logging.WARNING)
+
+
+logger = logging.getLogger(__name__)
+logger.propagate = True
+
+random.seed(47)  # for set_random_lsst_dds_partition_prefix
```
Comment (patrickingraham) [inline tests/auxtel/test_latiss_wep_align.py:59]: I don't think this is needed anymore?

File: tests/auxtel/test_latiss_wep_align.py [hunk 22]
```diff
@@ -0,0 +1,676 @@
...
+# Check to see if the test data is accessible for testing at NCSA
```
Comment (patrickingraham) [inline tests/auxtel/test_latiss_wep_align.py:60]: remove "at NCSA"

File: tests/auxtel/test_latiss_wep_align.py [hunk 23]
```diff
@@ -0,0 +1,676 @@
...
+# Depending upon how we load test data this will change
+# for now just checking if butler is instantiated with
+# the default path used on the summit and on the NTS.
+
+# DATAPATH = "/readonly/repo/main" # NTS - obsolete
+DATAPATH = "/repo/LATISS"  # SUMMIT (warning!)
+# DATAPATH = "/repo/LATISS" # TTS (TBC)
```
Comment (patrickingraham) [inline tests/auxtel/test_latiss_wep_align.py:67]: you ran these tests on the TTS right? I think the datapath is the same so you can remove line 67 and make this SUMMIT and TTS

File: tests/auxtel/test_latiss_wep_align.py [hunk 24]
```diff
@@ -0,0 +1,676 @@
...
+
+try:
+    butler = dafButler.Butler(
+        DATAPATH,
+        instrument="LATISS",
+        collections=["LATISS/raw/all", "LATISS_test_data"],
+    )
+    DATA_AVAILABLE = True
+except FileNotFoundError:
+    logger.warning("Data unavailable, certain tests will be skipped")
```
Comment (patrickingraham) [inline tests/auxtel/test_latiss_wep_align.py:77]: why does this use `logger.warnings` whereas line 38 uses `warning.warn`

**Review by patrickingraham — COMMENTED** [review]
Looks good.
Did you run the tests on the TTS with the data?

**Comment by tribeiro** [comment]
> Looks good.
> Did you run the tests on the TTS with the data?

I did, the wep one is not passing because the results are different. I tried to figure out why and found the problem with the obscuration parameter. But even using the same value for that the results are still different. Not too different, but different enough that is course the checks to fail, given the threshold you set.

File: python/lsst/ts/externalscripts/auxtel/latiss_wep_align.py (hunk 18 above)
Comment (tribeiro) [reply python/lsst/ts/externalscripts/auxtel/latiss_wep_align.py:201]: This is running on a separate process, so it does not have access to the properties of the class. I am not sure what that collection is as well.

File: tests/auxtel/test_latiss_wep_align.py (hunk 24 above)
Comment (tribeiro) [reply tests/auxtel/test_latiss_wep_align.py:77]: I don't know.. I just copied what was in the `test_latiss_cwfs_align.py` file... I think you did that... I actually didn't want to copy and paste it but I didn't find a good way to not duplicate this test... :(

File: tests/auxtel/test_latiss_wep_align.py (hunk 23 above)
Comment (tribeiro) [reply tests/auxtel/test_latiss_wep_align.py:67]: Yes, I ran them on TTS.. All these came from the original file.. I'll cleanup both..

File: python/lsst/ts/externalscripts/auxtel/latiss_wep_align.py [hunk 25]
```diff
@@ -0,0 +1,286 @@
...
+    ra, dec = wcs.pixelToSkyArray(
+        result.brightestObjCentroidCofM[0],
+        result.brightestObjCentroidCofM[1],
+        degrees=False,
+    )
+
+    donut_catalog = pandas.DataFrame([])
+    donut_catalog["coord_ra"] = ra
+    donut_catalog["coord_dec"] = dec
+    donut_catalog["centroid_x"] = [result.brightestObjCentroidCofM[0]]
```
[the remark this answers is not in the document]
Comment (tribeiro) [reply python/lsst/ts/externalscripts/auxtel/latiss_wep_align.py:238]: yes.. I'm after doing some tests I decided that it is better to continue using best effort isr and `QuickFrameMeasurementTask` to do pre-process the data and find the donut source. I think we will have to have a conversation about what we want to do with the MT but we have tested best effort isr and quick frame measurement to exhaustion on AuxTel so, I am not ready to gamble on this.

File: python/lsst/ts/externalscripts/auxtel/latiss_wep_align.py (hunk 17 above)
Comment (tribeiro) [reply python/lsst/ts/externalscripts/auxtel/latiss_wep_align.py:180]: well.. not really.. as I mentioned below (sorry I am going through these from the bottom up), I am still using best effort isr and quick frame measurement to pre-process the data and find the donuts so, in a sense, this is still needed. 

I think we can have a conversation about whether we should stop using best effort  isr and quick frame measurement, but we spend quite some time improving these so, I think we should continue using them.

File: tests/auxtel/test_latiss_wep_align.py (hunk 24 above)
Comment (patrickingraham) [reply tests/auxtel/test_latiss_wep_align.py]: Does it matter? I honestly didn't know there were two different "types" of warnings...

Comment (tribeiro) [reply tests/auxtel/test_latiss_wep_align.py]: I think logging and warning have slightly different use-cases.
 
According to the [logging doc page](https://docs.python.org/3/howto/logging.html#when-to-use-logging):

> [warnings.warn()](https://docs.python.org/3/library/warnings.html#warnings.warn) in library code if the issue is avoidable and the client application should be modified to eliminate the warning
>
> [logging.warning()](https://docs.python.org/3/library/logging.html#logging.warning) if there is nothing the client application can do about the situation, but the event should still be noted

I think for this particular one the `warning.warn` is more suitable than the `logging.warning`.

I already modified the code to use `warning.warn`.

File: python/lsst/ts/externalscripts/auxtel/latiss_wep_align.py (hunk 17 above)
Comment (patrickingraham) [reply python/lsst/ts/externalscripts/auxtel/latiss_wep_align.py:181]: Ok. Let's have a chat but we can make changes under a different PR.

**Review by patrickingraham — APPROVED** [review]

**Comment by tribeiro** [comment]
I think we need to spend some tie trying to figure out why the results are different in the first place. I tried to figure it out when I was writing the tests but couldn't. One of the reasons I didn't continued with the effort was discovering that the value for the obscuration we're using for cwfs is wrong. I think we need to settle on the correct value before doing additional work.
